In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("../data/raw/food.csv")

In [3]:
nutrition = df[[
    "Category",
    "Description",
    "Data.Kilocalories",
    "Data.Protein",
    "Data.Carbohydrate",
    "Data.Fat.Total Lipid",
    "Data.Fiber",
    "Data.Sugar Total"
]].copy()

nutrition.head()

,Category,Description,Data.Kilocalories,Data.Protein,Data.Carbohydrate,Data.Fat.Total Lipid,Data.Fiber,Data.Sugar Total
0,BUTTER,"BUTTER,WITH SALT",717,0.85,0.06,81.11,0.0,0.06
1,BUTTER,"BUTTER,WHIPPED,WITH SALT",717,0.85,0.06,81.11,0.0,0.06
2,BUTTER OIL,"BUTTER OIL,ANHYDROUS",876,0.28,0.00,99.48,0.0,0.00
3,CHEESE,"CHEESE,BLUE",353,21.40,2.34,28.74,0.0,0.50
4,CHEESE,"CHEESE,BRICK",371,23.24,2.79,29.68,0.0,0.51


In [4]:
nutrition = nutrition.rename(
    columns={
        "Category": "category",
        "Description": "food",
        "Data.Kilocalories": "calories",
        "Data.Protein": "protein",
        "Data.Carbohydrate": "carbs",
        "Data.Fat.Total Lipid": "fat",
        "Data.Fiber": "fiber",
        "Data.Sugar Total": "sugar"
    }
)

nutrition.head()

,category,food,calories,protein,carbs,fat,fiber,sugar
0,BUTTER,"BUTTER,WITH SALT",717,0.85,0.06,81.11,0.0,0.06
1,BUTTER,"BUTTER,WHIPPED,WITH SALT",717,0.85,0.06,81.11,0.0,0.06
2,BUTTER OIL,"BUTTER OIL,ANHYDROUS",876,0.28,0.00,99.48,0.0,0.00
3,CHEESE,"CHEESE,BLUE",353,21.40,2.34,28.74,0.0,0.50
4,CHEESE,"CHEESE,BRICK",371,23.24,2.79,29.68,0.0,0.51


In [5]:
nutrition.isnull().sum()

category    0
food        0
calories    0
protein     0
carbs       0
fat         0
fiber       0
sugar       0
dtype: int64

In [6]:
nutrition.info()

<class 'pandas.DataFrame'>
RangeIndex: 7413 entries, 0 to 7412
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   category  7413 non-null   str    
 1   food      7413 non-null   str    
 2   calories  7413 non-null   int64  
 3   protein   7413 non-null   float64
 4   carbs     7413 non-null   float64
 5   fat       7413 non-null   float64
 6   fiber     7413 non-null   float64
 7   sugar     7413 non-null   float64
dtypes: float64(5), int64(1), str(2)
memory usage: 463.4 KB


In [7]:
nutrition.shape


(7413, 8)

In [8]:
nutrition.drop_duplicates()

,category,food,calories,protein,carbs,fat,fiber,sugar
0,BUTTER,"BUTTER,WITH SALT",717,0.85,0.06,81.11,0.0,0.06
1,BUTTER,"BUTTER,WHIPPED,WITH SALT",717,0.85,0.06,81.11,0.0,0.06
2,BUTTER OIL,"BUTTER OIL,ANHYDROUS",876,0.28,0.00,99.48,0.0,0.00
3,CHEESE,"CHEESE,BLUE",353,21.40,2.34,28.74,0.0,0.50
4,CHEESE,"CHEESE,BRICK",371,23.24,2.79,29.68,0.0,0.51
...,...,...,...,...,...,...,...,...
7408,FROG LEGS,"FROG LEGS,RAW",73,16.40,0.00,0.30,0.0,0.00
7409,MACKEREL,"MACKEREL,SALTED",305,18.50,0.00,25.10,0.0,0.00
7410,SCALLOP,"SCALLOP,(BAY&SEA),CKD,STMD",112,23.20,0.00,1.40,0.0,0.00
7411,SNAIL,"SNAIL,RAW",90,16.10,2.00,1.40,0.0,0.00


In [9]:
nutrition.shape

(7413, 8)

In [11]:
nutrition[
    nutrition["calories"] == 0
]

,category,food,calories,protein,carbs,fat,fiber,sugar
233,SALT,"SALT,TABLE",0,0.0,0.00,0.00,0.0,0.0
4093,CARBONATED BEV,"CARBONATED BEV,CLUB SODA",0,0.0,0.00,0.00,0.0,0.0
4106,CARBONATED BEV,"CARBONATED BEV,LO CAL,OTHER THAN COLA OR PEPPE...",0,0.1,0.00,0.00,0.0,0.0
4114,CARB BEV,"CARB BEV,LO CAL,OTHR THN COLA OR PEPPER,W/ ASP...",0,0.1,0.00,0.00,0.0,0.0
4123,CARBONATED BEV,"CARBONATED BEV,LO CAL,COLA OR PEPR,W/ NA SACC,...",0,0.0,0.10,0.00,0.0,0.0
4136,COFFEE,"COFFEE,BREWED FROM GROUNDS,PREP W/ TAP H2O,DEC...",0,0.1,0.00,0.00,0.0,0.0
4137,COFFEE,"COFFEE,BREWED,ESPRESSO,REST-PREP,DECAFFEINATED",0,0.1,0.00,0.18,0.0,0.0
4208,WATER,"WATER,BTLD,PERRIER",0,0.0,0.00,0.00,0.0,0.0
4209,WATER,"WATER,BTLD,POLAND SPRING",0,0.0,0.00,0.00,0.0,0.0
4217,WATER,"WATER,TAP,DRINKING",0,0.0,0.00,0.00,0.0,0.0


## Data Quality Observations

Foods with zero calories were investigated.

Many zero-calorie entries appear to be legitimate products such as water, coffee, and baking soda.

Some entries appear to be calculation-generated food records rather than individual foods and may require additional review before inclusion in meal-planning analyses.

Zero-calorie foods were retained in the cleaned dataset because they represent valid food records.

In [17]:
nutrition_clean = nutrition.copy()

In [13]:
# missing food names
nutrition[
    nutrition["food"].isna()
]

,category,food,calories,protein,carbs,fat,fiber,sugar


In [15]:
nutrition.duplicated().sum()

np.int64(1)

In [18]:
nutrition_clean[
    nutrition_clean["protein"] > 50
][[
    "food",
    "calories",
    "protein"
]].sort_values(
    "protein",
    ascending=False
)

,food,calories,protein
4797,"SOY PROT ISOLATE,K TYPE,CRUDE PROT BASIS",321,88.32
4700,"SOY PROT ISOLATE,PROT TECHNOLOGIES INTERNATION...",388,87.75
4701,"SOY PROT ISOLATE,PROT TECHNOLOGIES INTERNATION...",380,86.00
5858,"GELATINS,DRY PDR,UNSWTND",335,85.60
7057,"Seal, bearded (Oogruk), meat, dried (Alaska Na...",351,82.60
124,"EGG,WHITE,DRIED,PDR,GLUCOSE RED",376,82.40
158,"EGG,WHITE,DRIED",382,81.10
4660,SOY PROTEIN ISOLATE,338,80.69
4796,"SOY PROT ISOLATE,K TYPE",326,80.69
7138,"STEELHEAD TROUT,DRIED,FLESH (SHOSHONE BANNOCK)",382,77.27


## Dataset Limitations

The nutrition database contains many foods that are technically valid entries but would not commonly be used for grocery shopping or meal planning.

Examples include exotic dried animal products, specialty foods, and research-oriented foods.

Future versions of the project may create a filtered dataset focused on common grocery foods.

In [19]:
nutrition_clean[
    (nutrtion_clean["protein"] >= 20)
    &
    (nutrition_clean["calories"] <= 300)
][[
    "food",
    "calories",
    "protein"
]].sort_values(
    "protein",
    ascending=False
)

,food,calories,protein
4311,"COD,ATLANTIC,DRIED&SALTED",290,62.82
7133,"SEAL,BEARDED (OOGRUK),MEAT,PART DRIED (ALASKA ...",270,62.38
7120,"CARIBOU,SHLDR MEAT,DRIED (ALASKA NATIVE)",271,59.38
3299,"SEAWEED,SPIRULINA,DRIED",290,57.47
7076,"WALRUS,MEAT,DRIED (ALASKA NATIVE)",251,57.00
...,...,...,...
987,"TURKEY,YOUNG TOM,DK MEAT,MEAT ONLY,RAW",123,20.04
898,"TURKEY,LIVER,ALL CLASSES,CKD,SIMMRD",273,20.02
6650,"GREEN GIANT,HARVEST BURGER,ORIG FLAV,ALL VEG P...",153,20.00
182,"CHEESE,PARMESAN,DRY GRATED,RED FAT",265,20.00


## Business Context Observation

Raw protein rankings are dominated by dried meats, specialty foods, and uncommon food sources.

While nutritionally accurate, these foods are not representative of typical grocery shopping behavior.

Future analyses should focus on foods commonly available in grocery stores and meal planning contexts.

In [20]:
nutrition_clean[
    nutrition_clean["food"].str.contains(
        "CHICKEN|TURKEY|BEEF|SALMON|COD|TUNA|EGG|YOGURT|CHEESE",
        case=False,
        na=False
    )
][[
    "food",
    "calories",
    "protein"
]].sort_values(
    "protein",
    ascending=False
)

,food,calories,protein
124,"EGG,WHITE,DRIED,PDR,GLUCOSE RED",376,82.40
158,"EGG,WHITE,DRIED",382,81.10
123,"EGG,WHITE,DRIED,FLAKES,GLUCOSE RED",351,76.92
4311,"COD,ATLANTIC,DRIED&SALTED",290,62.82
7129,"FISH,SALMON,CHUM,DRIED (ALASKA NATIVE)",378,62.09
...,...,...,...
640,"FISH OIL,COD LIVER",902,0.00
644,"FISH OIL,SALMON",902,0.00
630,"FAT,TURKEY",900,0.00
611,"FAT,CHICKEN",900,0.00


## Protein Density Observation

Results remain dominated by dried foods even after filtering to common grocery categories.

Drying removes water content and artificially increases protein concentration by weight.

For meal-planning purposes, fresh foods may be more representative of typical consumer choices than dried products. 

Q. What foods would I realistically buy at a grocery store for a high-protein meal?

In [21]:
nutrition_clean[
    nutrition_clean["food"].str.contains(
        "CHICKEN",
        case=False,
        na=False
    )
][[
    "food",
    "calories",
    "protein"
]].sort_values(
    "protein",
    ascending=False
)

,food,calories,protein
783,"CHICKEN,BROILERS OR FRYERS,BREAST,MEAT ONLY,CK...",187,33.44
850,"CHICKEN,STEWING,LT MEAT,MEAT ONLY,CKD,STWD",213,33.04
760,"CHICKEN,BROILERS OR FRYERS,LT MEAT,MEAT ONLY,C...",192,32.82
741,"CHICKEN,BROILERS OR FRYERS,GIBLETS,CKD,FRIED",277,32.54
779,"CHICKEN,BROILERS OR FRYERS,BREAST,MEAT&SKN,CKD...",222,31.84
...,...,...,...
6545,"MCDONALD'S,MCCHICKEN SNDWCH (WITHOUT MAYONNAISE)",240,11.08
736,"CHICKEN,BROILERS OR FRYERS,SKN ONLY,CKD,FRIED,...",394,10.32
6729,"CHICKEN POT PIE,FRZ ENTREE,PREP",214,6.08
767,"CHICKEN,BROILERS OR FRYERS,FAT,RAW",629,3.73


In [22]:
nutrition_clean[
    nutrition_clean["food"].str.contains(
        "BEEF",
        case=False,
        na=False
    )
][[
    "food",
    "calories",
    "protein"
]].sort_values(
    "protein",
    ascending=False
)

,food,calories,protein
3842,"BEEF,RND,TOP RND,LN,0""FAT,ALL GRDS,CKD,BRSD",199,36.12
3844,"BEEF,RND,TOP RND,LN,0""FAT,SEL,CKD,BRSD",190,36.12
3843,"BEEF,RND,TOP RND,LN,0""FAT,CHOIC,CKD,BRSD",207,36.12
3841,"BEEF,RND,TOP RND,LN&FAT,0""FAT,SEL,CKD,BRSD",200,35.62
3840,"BEEF,RND,TOP RND,LN&FAT,0""FAT,CHOIC,CKD,BRSD",216,35.62
...,...,...,...
3764,"BEEF,VAR MEATS & BY-PRODUCTS,SUET,RAW",854,1.50
1534,"GRY, HNZ Home style beef gravy",39,1.09
4088,"BEEF BROTH&TOMATO JUC,CND",37,0.60
6944,"BEEF,GROUND BF CALCULATOR - USER SELECTS LEAN/...",0,0.00


In [23]:
nutrition_clean[
    nutrition_clean["food"].str.contains(
        "TUNA",
        case=False,
        na=False
    )
][[
    "food",
    "calories",
    "protein"
]].sort_values(
    "protein",
    ascending=False
)

,food,calories,protein
4510,"TUNA,YELLOWFIN,FRSH,CKD,DRY HEAT",139,29.97
4411,"TUNA,FRSH,BLUEFIN,CKD,DRY HEAT",184,29.91
4472,"TUNA,LT,CND IN OIL,WO/SALT,DRND SOL",198,29.13
4412,"TUNA,LT,CND IN OIL,DRND SOL",198,29.13
4509,"TUNA,SKIPJACK,FRSH,CKD,DRY HEAT",132,28.21
4474,"TUNA,WHITE,CND IN OIL,WO/SALT,DRND SOL",186,26.53
4415,"TUNA,WHITE,CND IN OIL,DRND SOL",186,26.53
4413,"TUNA,LT,CND IN H2O,DRND SOL",116,25.51
4473,"TUNA,LT,CND IN H2O,WO/SALT,DRND SOL",116,25.51
4475,"TUNA,WHITE,CND IN H2O,WO/SALT,DRND SOL",128,23.62


In [24]:
nutrition_clean[
    nutrition_clean["food"].str.contains(
        "SALMON",
        case=False,
        na=False
    )
][[
    "food",
    "calories",
    "protein"
]].sort_values(
    "protein",
    ascending=False
)

,food,calories,protein
7129,"FISH,SALMON,CHUM,DRIED (ALASKA NATIVE)",378,62.09
7147,"SALMON,RED (SOCKEYE),FILETS W/ SKN,SMOKED (ALA...",345,60.62
7067,"FISH,SALMON,KING,CHINOOK,SMOKED,BRINED (ALASKA...",430,39.90
7124,"FISH,SALMON,RED,(SOCKEYE),CND,SMOKED (ALASKA N...",206,35.19
7065,"FISH,SALMON,KING,CHINOOK,KIPPERED,CND (ALASKA ...",266,30.70
4375,"SALMON,COHO,WILD,CKD,MOIST HEAT",184,27.36
4379,"SALMON,SOCKEYE,CKD,DRY HEAT",216,27.31
7117,"Fish, salmon, red, canned, bones removed (Alas...",161,27.31
4500,"SALMON,CHUM,CKD,DRY HEAT",154,25.82
4499,"SALMON,CHINOOK,CKD,DRY HEAT",231,25.72


## Open Question

the dataset contains multiple preparations of the same food (raw, cooked, dried, frozen, etc).

For meal-planning purposes, cooked foods may provide the most realistic estimate of nutrition consumed by the user.

Future iterations should standardize food preperation methods. 

In [25]:
nutrition_clean[
    nutrition_clean["food"].str.contains(
        "CHICKEN",
        case=False,
        na=False
    )
][[
    "food",
    "calories",
    "protein"
]]

,food,calories,protein
611,"FAT,CHICKEN",900,0.00
721,"CHICKEN,BROILERS OR FRYERS,MEAT&SKN&GIBLETS&NE...",213,18.33
722,"CHICKEN,BROILER OR FRYER,MEAT&SKN&GIBLETS&NECK...",291,22.84
723,"CHICKEN,BROILERS OR FRYERS,MEAT&SKN&GIBLETS&NE...",272,28.57
724,"CHICKEN,BROILERS OR FRYERS,MEAT&SKN&GIBLETS&NE...",234,26.78
...,...,...,...
6447,"BURGER KING, Chicken Tenders",289,17.25
6545,"MCDONALD'S,MCCHICKEN SNDWCH (WITHOUT MAYONNAISE)",240,11.08
6729,"CHICKEN POT PIE,FRZ ENTREE,PREP",214,6.08
7279,"CHICKEN,MEATLESS",224,23.64


## Data Structure Observation

The dataset contains many representations of the same ingredient.

For example, searching for "chicken" returned 183 records including:

- Raw ingredients
- Cooked preparations
- Frozen meals
- Restaurant foods
- Fast-food items
- Processed products

This suggests that future cleaning efforts may require classifying foods into meal-planning categories.

In [27]:
nutrition_clean[
    nutrition_clean["food"].str.contains(
        "CHICKEN",
        case=False,
        na=False
    )
]["category"].value_counts()

category
CHICKEN                   163
No Category                 5
CHICKEN BREAST TENDERS      3
CHICKEN NUGGETS             2
CHICKEN PATTY               2
FAT                         1
GRAVY                       1
PATE                        1
LOUIS RICH                  1
KRAFT                       1
BURGER KING                 1
MCDONALD'S                  1
CHICKEN POT PIE             1
Name: count, dtype: int64

## Category Quality Assessment

The Category field is inconsistent.

Examples include:

- Ingredient categories (CHICKEN)
- Product categories (CHICKEN NUGGETS)
- Prepared meals (CHICKEN POT PIE)
- Brands (KRAFT)
- Restaurants (BURGER KING, MCDONALD'S)

This limits the usefulness of Category for filtering and suggests the need for a custom meal-planning classification system.

In [28]:
nutrition_clean["category"].nunique()

1183

In [29]:
nutrition_clean["category"].value_counts().head(30)

category
BEEF                       457
PORK                       304
CEREALS RTE                291
BABYFOOD                   221
No Category                198
LAMB                       198
SOUP                       188
CHICKEN                    163
CEREALS                    132
FAST FOODS                 130
CANDIES                    129
TURKEY                     124
CAMPBELL SOUP COMPANY      107
BEANS                      106
VEAL                        86
MCDONALD'S                  82
CHEESE                      73
POTATOES                    72
COOKIES                     63
OIL                         62
ALCOHOLIC BEV               62
SALAD DRSNG                 60
GAME MEAT                   56
BREAD                       55
CAMPBELL                    50
PUDDINGS                    48
SQUASH                      45
CAMPBELL'S CHUNKY SOUPS     42
MILK                        40
OSCAR MAYER                 40
Name: count, dtype: int64

## Category Field Assessment

The category field contains 1,183 unique values.

Categories are not standardized and include:

- Food groups (BEEF, PORK, CHICKEN)
- Prepared foods (SOUP, COOKIES)
- Restaurants (MCDONALD'S)
- Brands (CAMPBELL, OSCAR MAYER)
- Missing values (No Category)

Because category values represent multiple concepts, the field is not suitable for direct meal-planning classification without additional cleaning.